src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs

```diff
@@ -1,46 +1,61 @@
 using System;
+using System.Collections.Generic;
 using System.Collections.ObjectModel;
+using System.Diagnostics;
 using System.IO;
 using System.Linq;
 using System.Text.RegularExpressions;
+using RoslynPad.UI.Services;
+using RoslynPad.UI.Utilities;
 using RoslynPad.Utilities;
 
 namespace RoslynPad.UI
 {
+    [DebuggerDisplay("{Name}:{IsFolder}")]
     public class DocumentViewModel : NotificationObject
     {
+        private readonly IDisposable _documentFileWatcherDisposable;
+        private readonly DocumentFileWatcher _documentFileWatcher;
         internal const string DefaultFileExtension = ".csx";
         internal const string AutoSaveSuffix = ".autosave";
 
         private ObservableCollection<DocumentViewModel> _children;
         private bool _isExpanded;
         private bool? _isAutoSaveOnly;
         private bool _isSearchMatch;
+        private string _path;
+        private string _name;
 
-        private DocumentViewModel(string rootPath)
+        private DocumentViewModel(string rootPath, DocumentFileWatcher documentFileWatcher)
         {
+            _documentFileWatcher = documentFileWatcher;
+            _documentFileWatcherDisposable = _documentFileWatcher.Subscribe(OnDocumentFileChanged);
             Path = rootPath;
-            IOUtilities.PerformIO(() => Directory.CreateDirectory(Path));
-            IsFolder = true;
-            IsSearchMatch = true;
-        }
-
-        private DocumentViewModel(string path, bool isFolder)
-        {
-            Path = path;
-            IsFolder = isFolder;
-            Name = isFolder ? System.IO.Path.GetFileName(Path) : System.IO.Path.GetFileNameWithoutExtension(Path);
-            // ReSharper disable once PossibleNullReferenceException
+            IsFolder = IOUtilities.IsDirectory(Path);
+            Name = System.IO.Path.GetFileName(Path);
             IsAutoSave = Name.EndsWith(AutoSaveSuffix, StringComparison.OrdinalIgnoreCase);
             if (IsAutoSave)
             {
                 Name = Name.Substring(0, Name.Length - AutoSaveSuffix.Length);
             }
-
             IsSearchMatch = true;
         }
 
-        public string Path { get; set; }
+        public string Path
+        {
+            get => _path;
+            set
+            {
+                if (IsFolder)
+                {
+                    foreach (var child in Children)
```

Comment:
This will **recursively initialize all `Children` from the ctor** - a thing we try to avoid. Normally `Children` is evaluated lazily by the UI. It would be better to leave a simple setter, and add a method that does the recursion if `_children != null` (better make a `bool IsChildrenInitialized => _children != null` and it use).